### PR TITLE
Specifications & Corrections of PAL languages in GameIndex.

### PR DESCRIPTION
**Pull request by pgert**
Specifications & Corrections of PAL languages in GameIndex.
This PR is "ported" from PR #661.

### CODE REVIEW

**Comment by bositman** [comment]
You did something wrong...2,068 additions, 2,063 deletions not shown

**Comment by ssakash** [comment]
@bositman 

Github doesn't support viewing comparisons between master and the commit for files with large amount of differences. you'd have to get the raw file of the branch's GameIndex.dbf from here : https://github.com/pgert/pcsx2/blob/master/bin/GameIndex.dbf 

and finally you'd have to use any diff checker tool by comparing to the master's GameIndex.dbf to view the commit changes.

**Comment by willkuer** [comment]
I guess he knows that. The point is probably that pgert made 2000 changes and only 5 additions. I haven't checked the diff but possibly pgert inserted wrong line feeds or whitespaces in 2000 lines. I guess bositman just wants to keep git noise low.

**Comment by bositman** [comment]
Yeah, the whole file seems changed while there are only a handful of changes.

**Comment by pgert** [comment]
@bositman:

I missed removing this addition (which is responsible for 4 new lines):
Serial = SLES-53503
Name   = Star Wars - Battlefront II
Region = PAL-G

The 5th new line is the comment in this entry:
Serial = SLES-54454
Name   = Disgaea 2: Cursed Memories
Region = PAL-E
// Voices can be switched between Japanese and English via Options menu.

I hope it's OK.

**Comment by AdmiralCurtiss** [comment]
There are actually 2000+ changes in this file, mostly along the lines of this stuff:

[Changes 1](http://i.imgur.com/4YZsXQT.png) [Changes 2](http://i.imgur.com/u3QILEA.png)

Most of it seems reasonable to me, though I don't understand why you sometimes write M3 and sometimes write E-G-I or whatever.

**Comment by avih** [comment]
Could you please summarize _all_ the region names and what each of them mean?

**Comment by mirhl** [comment]
For the third *unanswered* time, languages and regions are whole [different](https://github.com/PCSX2/pcsx2/pull/661#issuecomment-138939236) things.

**Comment by avih** [comment]
> For the third unanswered time, languages and regions are whole different things. 

That's fine. So this patch adds several new language codes, and it would make a lot of sense to list all the codes and what each code means.

**Comment by pgert** [comment]
@ avih:
"Could you please summarize all the region names and what each of them mean?"

E  - English
G  - German
I  - Italian
S  - Spanish
F  - French
P  - Portuguese
R  - Russian
DU - Dutch
NO - Norwegian
SE - Swedish
DA - Danish
FI - Finnish
CZ - Czech
PL - Polish
HR - Croatian
HI - Hindu

**Comment by avih** [comment]
@pgert this is great!

However, your list doesn't cover all the changes which your patch does. E.g. it also changes from `PAL-Unk` to `PAL-M3` or `PAL-M5` or `PAL-M11` etc, and these M3/M5/etc do not appear at your list.

Please complete the list with all the existing/available codes, or if you don't know all of them, only list the ones which your patch uses.

After that, please update the commit message of your commit (which is _not_ the github discussion comments nor the github PR title - hint, you'll have to `commit --amend` and then `push -f` to push the new nessage) to convey this useful info. Something like:
```
This fixes the language codes, mostly from PAL-Unk to e.g. PAL-CZ or PAL-M5, etc.

Here is the full list of codes:
M5 - <whatever>
E - English
G - German
...
```

**Comment by yorkiesandskittles** [comment]
Wouldn't this system be confusing in literally every other language?

**Comment by bositman** [comment]
M3/5/11 are multilanguage with 3/5/11 languages respectively

**Comment by avih** [comment]
@bositman I know what some of them are, my point was to summarize the list of all available codes in one place - where it matters - at the commit message which changes them.

**Comment by mirhl** [comment]
@avih
*Regions* are only NTSC-U/C, NTSC-J, NTSC-C and PAL that's what I was saying.

**Comment by avih** [comment]
@mirhl sorry, I read what you're saying, but I'm not sure whether you're implying something or not. E.g. do you imply that this PR is not good for some reason? or that the changes should be according to a different standard? or anything else?

**Comment by mirhl** [comment]
AdmiralCurtiss idea seems way more logical. 
*Regions* are *regions*, aka those things that lock your games out of foreigner consoles. 
Localizations may be a good way to further categorize games then, but it doesn't make sense to tag them with "M3" or "M5", given there are probably multiple game releases with the same *number* of languages, but different ones.

**Comment by avih** [comment]
I don't see an idea on this PR, but there could be an implied objection hidden at "Regions are regions".

Spell it out for me please:
1. Do you think the regions/languages at the dbf file should be changed?
2. If yes, how would you do it differently than this PR?
3. If yes, how?

**Comment by mirhl** [comment]
[This](http://imgur.com/gallery/OGyYCwC/) is what I mean. 
#661#issuecomment-138939236 is what I was looking to.

**Comment by avih** [comment]
Well, the serial "defines" both the region and the language (in the sense that a serial is not distributed in different regions or with different languages AFAIK), so this whole region/language info is fully redundant and PCSX2 itself doesn't make any use of it - It's mainly for convenience when browsing the db or for displaying at the console title, that's it.

So with that in mind, it doesn't really matter what kind of info we put there IMO.

Personally I'd never have touched it manually and I'd have recommended to others not to either, but pgert decided he wants to.

Now, please describe exactly what's the problem with this PR. Please don't refer me to comments made on other places etc. Spell it out for me exactly: This PR is not good because: A, B, C.

And then let's take it from there.

**Comment by mirhl** [comment]
Yeah, I know it's just for convenience (in easier discerning the games, right?). 
So I don't see the point with M*x* designations, if then your risk multiple games releases to appear identical (except serial of course)
 ---------------------------------------------
Serial   = SCES-50916
Name     = Ratchet & Clank
Region   = PAL
Language = English
Language = French
Language = German
Language = Italian
Language = Spanish
Compat   = 5
 --------------------------------------------- 
This was the plan you could find in the other discussion, which seems less prone to "duplicates". 
Btw, while I was looking for info on the matter, I realized we even lack [games](https://cloud.githubusercontent.com/assets/9366725/9901682/5e2767d4-5c69-11e5-88a0-99dcc40b736e.jpg) data for the whole chinese region.

**Comment by avih** [comment]
So basically what you've been saying this whole time is that the M<N> descriptions are not unique and that there could be e.g. two serials which are both PAL-M5 but with different set of languages.

And the solution you suggest is to change the format and add all languages explicitly, e.g. by adding extra lines where each language has its own line.

If there are no issues other than that, then IMO we don't need these extra details.

Again, the region and language are not used by anything within PCSX2, and it's just a convenience thing and _some_ extra info when browsing the DB. If you've ripped a DVD, you'll know it's PAL (and even if you don't, it doesn't matter) and surely you don't need PCSX2 to tell you that this disk has such and such languages, right? It's completely not helpful, not useful, and not used by PCSX2.

So IMO having either a single language or an M<N> description which says that this DVD has N different languages is good enough (I'd go further and say that the language is not required at all, but whatever, we already have some languages at the DB and there's no point in removing them).

So in this sense I disagree that we should make it more detailed than what this PR does.

I think such change should also be compatible with NTSC languages, if there are such. E.g. let's say that there are NTSC-U DVDs with either English or Spanish languages, then we could use NTSC-U-E and NTSC-U-S. I really don't want to go there, but theoretically we could.

However, I might have an issue with the codes which this PR uses for the languages.

@pgert where did you take these codes from? are they a Sony thing? or did you just looked them up at the internet and found two-letter codes for the missing languages?

And if we use two letter codes for some, shouldn't we use two letter codes for all? e.g. PAL-EN instead of PAL-E?

So if these letter are official Sony codes, then this PR is fine for me. Otherwise, I'd want to see a good reason to have some languages described by one letter code and other described by two letter codes.

**Comment by pgert** [comment]
"So basically what you've been saying this whole time is that the M descriptions are not unique and that there could be e.g. two serials which are both PAL-M5 but with different set of languages."
Reply:
Well for PAL-M2 it's quite common, but I doubt it's so with any PAL-M5.


" It's completely not helpful, not useful, and not used by PCSX2"
Reply:
It's primarily for presentation, but it could be useful (I imagine) for porting patches from one disc to another.


"where did you take these codes from?"
Reply:
As I specified in PR #661 they are mainly extracted from:
http://redump.org/discs/system/ps2/
http://ns348841.ip-91-121-109.eu/psxdata/psx2/pal_list2.html
http://wiki.redump.org/index.php?title=Sony_Playstation_2_PAL_Missing_Dumps


"And if we use two letter codes for some, shouldn't we use two letter codes for all? e.g. PAL-EN instead of PAL-E?"
Reply:
Generally only one letter is used for the more common languages; english etc - Yes?


"So if these letter are official Sony codes, then this PR is fine for me."
Reply:
Don't know if they are official codes, but they are the usually used codes.


What this PR primarily does is to sort out the PAL-E mess.
Some have used PAL-E for Englsh, others for Europa (PAL-M5), and others still for España.
With this PR all PAL-E should refer to English, while the ones refering to Europa have PAL-M5 or something as a code, and the ones refering to España have PAL-S as a code.

**Comment by avih** [comment]
> What this PR primarily does is to sort out the PAL-E mess.
> Some have used PAL-E for Englsh, others for Europa (PAL-M5), and others still for España.
> With this PR all PAL-E should refer to English, while the ones refering to Europa have PAL-M5 or something as a code, and the ones refering to España have PAL-S as a code.

and

> they [the codes] are mainly extracted from:
> http://redump.org/discs/system/ps2/
> http://ns348841.ip-91-121-109.eu/psxdata/psx2/pal_list2.html
> http://wiki.redump.org/index.php?title=Sony_Playstation_2_PAL_Missing_Dumps

and

> Generally only one letter is used for the more common languages; english etc

All this is great info, and it's exactly the kind of info which should be part of the commit message.

Now, please update the **commit message** with:
- The 3 types of info I quoted above (copying them verbatim from the quote is OK).
- The full list of language codes you used (the one you listed earlier).
- A short (a line or two at most) explanation of M2/M5/etc

The reason I'm not terribly excited about this PR in general is that on one hand it's nice info to have, but OTOH it's not really required and on the third(!) hand, it might break some tools which handle the games db (I can think of PCSX2 itself which loads the db, the db editor which is currently only enabled in debug builds, and the db compatibility update tool at the `tools/GameIndex.dbf-tool`).

The critical one is of course PCSX2 itself which must not break by this update. I don't think it would, but this _must_ be tested before we merge this PR. @pgert have you tested it? if no, please do and report back here.

Let's hope the other tools won't get broken by it and that if they do, someone will be able to fix them.

**Comment by mirhl** [comment]
> And the solution you suggest is to change the format and add all languages explicitly, e.g. by adding extra lines where each language has its own line.

Extra lines, or a single line with languages separated by commas, yes. 
> If there are no issues other than that, then IMO we don't need these extra details.

There are not other issues other than that, indeed, but I don't see why the *number* of languages would be informative of anything while the exact languages would not. 

> let's say that there are NTSC-U DVDs with either English or Spanish languages, then we could use NTSC-U-E and NTSC-U-S.

This is completely silly imo. Or at least *personally* I wouldn't be able to understand whatever that means. 
Btw the region code is NTSC U/C.

**Comment by avih** [comment]
> I don't see why the number of languages would be informative of anything while the exact languages would not.

I never said it would not be informative. Just that it gives more details compared to e.g. M5, and we can live fine without these extra details.

**Comment by mirhl** [comment]
Ok, I think I got it now. This is only about expanding the current odd situation with region being the actual region + number of languages attached. 

Could somebody enlighten me on any question, doubt or wonder the information regarding the number of languages could address?

**Comment by avih** [comment]
> Could somebody enlighten me on any question, doubt or wonder the information regarding the number of languages could address?

Curiosity.

**Comment by mirhl** [comment]
Ok, so after a whole day it seems there's no purpose for this kind of data, other than.. well whatever. 
Since this commit is about giving a batter layout to the current one used since the [beginning](https://github.com/PCSX2/pcsx2/commit/2a05b6d44951a93f2e39354e7e1dbe91b35be254) I won't have anything else to say. 

But imo, it wasn't worth all the time spent, considering it's indeed *more* than the bare essential, but still not enough to provide any appreciable addition to that.

**Comment by pgert** [comment]
@avih:
I don't know how to squash commits with GitHub, so I have to close this PR and open a new one.

**Comment by avih** [comment]
@pgert thanks, the new message looks very good.

Also, It's actually really better to NOT open a new PR, since all the discussion happened on this PR, and a new one will lose all the context.

Just learn how to do that.

I can squash it for you (though I'll only do it once), or you can learn how to squash and we'll wait till you do it yourself. It's not like we're in a hurry, and squashing is very useful in general. So come on, start reading ;)

And please don't ope a new PR.

**Comment by pgert** [comment]
Sorry, too late.
Ported to PR #835.
I can't reverse it.
The link in the new PR to this one should be fine enough.

**Comment by avih** [comment]
Notice how this PR is also "ported" from #661.

You can't just open a new PR for every change you do.

If you have enough patience to change 2000 PAL language codes, you should be able to learn how to squash as well.

Now do it.

**Comment by micove** [comment]
I think he can't do it anymore even if he wanted to. I'm not sure but I think he went the extra mile and deleted the old repository and just reforked it to avoid rebasing/squashing. Notice how it says 

"pgert wants to merge 2 commits into PCSX2:master from unknown repository"

I think it's unknown because it was deleted.

**Comment by AdmiralCurtiss** [comment]
For the future, here's the easy way to squash: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

I'd also recommend not committing to master and instead creating branches for every PR.

**Comment by avih** [comment]
> The link in the new PR to this one should be fine enough.

It's not. The same as the link in this PR to the previous one isn't.

> I think he can't do it anymore even if he wanted to.

Possibly. I've pushed it myself as 9eb5f36

@pgert please learn how to squash and how to (force) push changes to PRs, and don't "port" PRs due to minor changes, it's really hard to follow the discussion this way. Changing code or messages after some review is part of the workflow. Opening a new PR for any such change breaks the workflow.

**Comment by micove** [comment]
If the interactive rebase was too "hard" then it could have been done as:
git reset --soft HEAD~2
git commit -c ORIG_HEAD
git push -f

Just moving the head back and using the commit message from the orignal head.

**Comment by avih** [comment]
> If the interactive rebase was too "hard"

This works for a limited set of scenarios. He should probably get comfortable with squashes in general. Also, I don't know what's his environment. It could be some GUI tool or github itself, and each environment has it's own way of doing things.

There are a lot of guides and helpful people, including here and including you and myself. If he's asked to do something and he doesn't know how, he should ask for help rather than decide on his own to do something else.

I lost my patience here since it's not the first time it happens with him.

**Comment by avih** [comment]
@pgert I examined the new file, and I noticed it has 107 games with `PAL-Unk`, 31  `PAL-F-G`, 27 `PAL-E-G`, couple of `PAL-E-F-G` and many more which are neither a single language nor have codes of `PAL-M<N>`.

While the quest to sort out the `PAL-E` codes was a noble one, it does end up with inconsistency (maybe less inconsistency than before, but still).

What do you recommend to do with those?

**Comment by pgert** [comment]
@ AdmiralCurtiss:
Thanx for the link; i will look in to it.

@ avih:
Nothing really.
PAL-F-G says more than PAL-M2, so why not keep it.
If someone is up for it then it's fine by me for them to look up what languages is supported, and turn the PAL-M2's into PAL-X-Y through a PR.
I have no improvement-suggestion for the PAL-Unk entries, or the PAL-Unk code.

**Comment by avih** [comment]
Sounds good to me. Thanks.
